.github/workflows/pr-slack-notification.yaml

```diff
@@ -0,0 +1,304 @@
+          # Load previous PR data if exists
+          if [ -f cache/notified_prs.json ]; then
+            previous_prs=$(cat cache/notified_prs.json)
+          else
+            previous_prs="[]"
+          fi
+
+          pr_list=""
+          new_notified_prs="[]"
+          notify=false
+
+          for pr_info in $(echo "$pr_infos" | jq -r '.[] | @base64'); do
+            _jq() {
+              echo "$pr_info" | base64 --decode | jq -r "${1}"
+            }
+
+            pr_number=$(_jq '.number')
+            pr_title=$(_jq '.title')
+            pr_user=$(_jq '.user.login')
+            pr_url=$(_jq '.html_url')
+            pr_draft=$(_jq '.draft')
+            pr_created_at=$(_jq '.created_at')
+            pr_updated_at=$(_jq '.updated_at')
+
+            pr_data=$(jq -n --arg number "$pr_number" --arg updated_at "$pr_updated_at" '{number: $number, updated_at: $updated_at}')
+            new_notified_prs=$(echo "$new_notified_prs" | jq --argjson pr_data "$pr_data" '. += [$pr_data]')
+
+            echo "New Notified PR List------ $new_notified_prs"
+
+            # Check if the PR is new or updated
+            previous_pr=$(echo "$previous_prs" | jq --arg number "$pr_number" '.[] | select(.number == $number)')
+            echo "Checking PR is new or not --- $previous_pr"
+            if [ -z "$previous_pr" ] || [ "$(echo "$previous_pr" | jq -r '.updated_at')" != "$pr_updated_at" ]; then
+              draft_status=""
+              echo "------Draft value ---- $pr_draft"
+              if [ "$pr_draft" = "true" ]; then
+                draft_status="\n*Draft*: true"
+              fi
+
+              pr_list="${pr_list}\n*Pull Request* #${pr_number}: ${pr_title}\n*Created by*: ${pr_user}\n*URL*: ${pr_url}${draft_status}\n*Created At*: ${pr_created_at}\n*Last Updated At*: ${pr_updated_at}\n"
+              notify=true
+              echo "Boolean Value for notify ------- $notify"
+            fi
+          done
+
+          # Save current PR data for future comparison
+          echo "$new_notified_prs" > cache/notified_prs.json
+
+          if [ "$notify" = true ]; then
+            echo -e "$pr_list" > pr_list.txt
+            echo "::set-output name=notify::true"
+          else
+            echo "::set-output name=notify::false"
+          fi
+
+      - name: Fetch closed pull requests since the last run
+        id: fetch_closed_prs
+        run: |
+          if [ -f cache/last_run_timestamp.txt ]; then
+            last_run_timestamp=$(cat cache/last_run_timestamp.txt)
+          else
+            last_run_timestamp=$(date -u -d "-24 hours" +%Y-%m-%dT%H:%M:%SZ)
+          fi
+
+          echo "Last run timestamp: $last_run_timestamp"
+
+          closed_prs=$(curl -s -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" \
+          "${{ env.API_URL }}?state=closed&sort=updated&direction=desc&per_page=100")
+
+          closed_pr_list=""
+
+          for pr_info in $(echo "$closed_prs" | jq -r '.[] | @base64'); do
```

Comment:
These lines of code are present in step "Fetch all opened pull request details using condition" and here. Is there is any way to avoid this code duplication?

Reply:
It's challenging to avoid code duplication by creating a reusable step or script that handles the common functionality between 'Fetch closed pull requests since the last run' and 'Fetch all opened pull request details using conditions' because both steps involve different logics. Additionally, we may require additional conditional checks, which complicates efforts to prevent code duplication.

Reply:
@aparnamichael , I simplified the code for the step `'Fetch closed pull requests since the last run'`, and I attempted the same approach to simplify the code for the step `'Fetch all opened pull request details using conditions'` as well. I avoided using a for loop and instead utilized jq with select queries to filter out values. However, when checking the conditions we implemented for listing PRs (specifically, `omitting PRs already listed in Slack or those that have been updated or are new`), it did not succeed. If necessary, I will investigate further.

Reply:
@vaisakhkannan Thank you for simplifying the code where able. If it is becoming too complicated or too complex to simplify the code elsewhere, there is no need to do so.